### ISSUE TITLE
HMR not working on clean install - any change causes full page reload

### ISSUE THREAD

**Issue by PeterDraex** [withastro/astro#8916]
### Astro Info

```block
Astro                    v3.3.4
Node                     v20.8.1
System                   Linux (x64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             none
```

### If this issue only occurs in one browser, which browser is a problem?

Chrome, Edge

### Describe the Bug

HMR is broken in v3.3.4. Any change causes a full page reload. Reproducible on my local machine and on StackBlitz.

Steps to reproduce:
1. `npm create astro@latest`, choose Typescript-strict, install dependencies
2. Open web browser, Welcome To Astro page loads ok
3. Open Developer tools, Network tab
4. Open `src/Components/Card.astro` and change row 14 from `{title}` to `X{title}`, save
5. Full page reload was performed. You can see this in the Network tab of Developer tools - all files were loaded again, including the favicon :).

![image](https://github.com/withastro/astro/assets/18402223/f51c5332-c983-405e-9f23-29f8f57386fb)

### What's the expected result?

After modifying Card component, only `src/Components/Card.astro` is reloaded. State of the page is not reset. Existing websocket connection is preserved. 

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-3xrkdv?file=src%2Fcomponents%2FCard.astro

### Participation

- [ ] I am willing to submit a pull request for this issue.
```[tasklist]
### Tasks
```

**Comment by alex-sherwin** [comment]
Observing same problem... no matter what I do the Vite HMR is always pushing down the following message over the HMR websocket:

```json
{type: "full-reload"}
```

Which, of course, causes a full page refresh

**Comment by natemoo-re** [comment]
I believe this is expected! Astro has never supported client-side (partial) HMR when an `.astro` file is updated. The page is always reloaded for markup changes.

It's a bit confusing because there are scenarios where client-side HMR can occur:
- When a CSS file is updated, that can be swapped on the client
- When a script is updated, that can be swapped on the client
- When a client-side component is updated, its `.js` file exists in the client and has a reference to its parent file. The updated module with new rendering logic can be swapped in on the client.

Unfortunately, since `.astro` files don't exist on the client, the client has no reference to understand which files to listen to for updates.

**Comment by PeterDraex** [comment]
@natemoo-re I think this is not expected, since [release notes](https://astro.build/blog/astro-021-preview/#hmr-meet-html) for v0.21 say this:

> Starting in v0.21.0, Astro will support full HMR for Astro components and pages. Change any .astro file in your codebase, and watch the dev server update the page without a full refresh and without losing any client state.

**Comment by natemoo-re** [comment]
@PeterDraex Those release notes are referring to an experiment that never shipped in a stable 1.0.0+ version of Astro. Unfortunately there were too many edge cases and interactions with other framework's HMR logic. We had to rollback that feature in favor of consistency and reliability.
